### ISSUE TITLE
Using `Array.smooth()` as a parameter causes various glitches during the first seconds

### ISSUE THREAD

**Issue by jacopo-salamina** [hydra-synth/hydra-synth#154]
Whenever a Hydra function uses an array as one of its parameters, and this array's `smooth` property is set to a non-zero value:

- the output buffer chosen for displaying this function will behave in an "unexpected way";
- this will only happen during the very first seconds Hydra is started, after which the output buffer will always behave as expected;
- the amount of time during which this behaviour can be observed is a function of:
  - the global variable `bpm`;
  - the array's properties `smooth`, `speed` and `offset`.

The "unexpected way" in question depends on the functions used. The three examples shown below include each the expected and actual behaviour observed (all three assume `bpm` is set to 30):

```javascript
/*
 * A white square should pop up immediately and start scrolling horizontally.
 * Instead, during the first second, the entire screen goes white.
 */
shape(4, 0.1).scroll([1, 0].smooth(), 0).out(o0)

/*
 * The screen should transition smoothly from red to black, and vice-versa.
 * Instead, during the first second, the screen stays black, then jumps to red; afterwards, it fades to/from black as expected.
 */
solid([1, 0].smooth(), 0, 0).out(o0)

/*
 * A white triangle should fade in and out, revealing and hiding a white square.
 * Instead, during the first second, the screen stays black, then the triangle becomes immediately visible; afterwards, it fades in and out as expected.
 */
shape(4, 0.1).blend(shape(3, 0.3), [1, 0].smooth()).out(o0)
```

This behaviour can only be observed during the first seconds Hydra is started. For example, when using the public web demo, this behaviour happens during the first seconds the tab is opened, and right after reloading it. Afterwards, the output behaves as expected, even if the input script is reloaded using Ctrl-Shift-Enter.
The amount of time during which the output glitches out (which I'll call `T` from now on, to avoid repetitions) only seems to depend on the properties listed above. In the three examples shown above, the unexpected behaviour lasts exactly one second. Altering any property as described below changes `T`:

- increasing the global variable `bpm` decreases `T` and vice-versa (e.g. in the examples above, setting `bpm` to 7.5 increases `T` to 4 seconds);
- increasing the array's property `smooth` increases `T` and vice-versa (e.g. in the examples above, calling `smooth(4)` on the arrays increases `T` to 4 seconds);
- increasing the array's property `speed` decreases `T` and vice-versa (e.g. in the examples above, calling `fast(0.25)` on the arrays increases `T` to 4 seconds);
- increasing the array's property `offset` decreases `T` and vice-versa (e.g. in the examples above, calling `offset(0.3)` on the arrays decreases `T` to roughly half a second).

If `ease()` is called on the array, `T` seems not to be affected by the new easing function chosen.
In a similar fashion, `T` seems not to be affected by the Hydra function in which the array is used. In the examples shown above, the unexpected behaviors always happen for one second.

This bug was tested on version 1.3.29, using both the public web demo and some local scripts (both were using the version provided by UNPKG, which, at the time of opening this issue, was 1.3.29). The browser used was Chromium 119, running on Arch Linux.

**Comment by geikha** [comment]
This is because Hydra tries to interpolate between `array[-1]` and `array[0]`, to put it in simple terms. We would have to decide if it should simply start one element further, if it should wait, if it should interpolate with the last element of any array... etc.

What do you (everyone reading) think would be the right choice?

**Comment by jacopo-salamina** [comment]
# TL;DR

We should interpolate with a fixed value (possibly the array's last value). Save for the very first seconds, the array would behave in the same way as it does now (no compatibility issues with existing code snippets). Also, using a fixed value during the first seconds, to me, doesn't play nicely with the overall idea behind `Array.smooth()`. If we're expecting a value to change smoothly and constantly, why should we get a fixed value at the start?

# Full text

Here are 4 possible options, each with my own personal reasons for choosing it or not. Feedback/corrections are more than welcome.

## Option 1: shift the internal indexes by 1

After debugging and inspecting `arrayUtils.getValue()`, this looks like the easiest behavior to implement: we can just add 1 to the offset computed internally.

However, since the shift in the internal indexes also affects the values computed later in time, this would be a major change for the current array behavior, with all sorts of compatibility issues with the existing scripts. I can already think of a scenario where a script's behavior would change drastically: any script which involves both a "smoothed array" (one whose property `smooth` is set to anything except 0) and either:

- a "non-smoothed array" (one whose property `smooth` is set to 0), or
- a lambda function which uses `time`.

Such a script would have its behavior changed because, like I said in the beginning, we'd introduce an implicit offset of 1 only to "smoothed arrays", while everything else would produce the same exact values at the same exact instants. The shift in time of the generated values wouldn't be consistent across the different arrays, lambdas, etc.

## Option 2: output `arr[0]` during the first seconds

The reason I'm only considering `arr[0]`, and not another fixed constant, is the following: after the first few seconds had passed, and the array started interpolating values as usual, there would be a sudden jump between the fixed constant and the result of the usual interpolation. This wouldn't play nicely with the idea behind `Array.smooth()`; the generated value is supposed to change "smoothly", after all.

Even if we used `arr[0]` for the initial values, I'm still unhappy with this option. The reason is that, while the sudden jump mentioned above is avoided, the array behaves in a substantially different way (it outputs a fixed value), compared to its behavior later in time (it interpolates between consecutive values).

One could argue that, when using any value between 0 and 1 for `smooth`, a similar situation happens: the first value of the array is used for a while, then there's the transition to the next one, then the latter is used for a while, etc. Now, I think such a counter-argument is also valid; however:

- using `smooth` to enforce this behavior, both in the short and the long term, is a choice under the end user's control;
- on the other hand, there would be no way of disabling the initial behavior introduced by option 2.

Overall, to me, the idea of using a fixed value during the first seconds clashes with the design behind `Array.smooth()`; however, I don't really have a strong argument against this option.

## Option 3: replace `arr[-1]` with the array's last value

I can't think of any argument against this option.

## Option 4: replace `arr[-1]` with a fixed constant

Before considering this option, we would have to come up with a very strong reason to choose a particular constant. Otherwise, any given choice could be deemed arbitrary. Having a strong argument for a specific constant becomes even more important when considering the end user won't be able to modify it.

**Comment by ojack** [comment]
I vote for option #3 (interpolate from last value) as well! As a user,  it is what I would expect to happen.

**Comment by ojack** [comment]
merged!
